File: practice-java/inventory.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

// Define the Product class
class Product {
    String productId;
    String productName;
    double price;

    Product(String productId, String productName, double price) {
        this.productId = productId;
        this.productName = productName;
        this.price = price;
    }

    void displayProductInfo() {
        System.out.println("Product ID: " + productId);
        System.out.println("Product Name: " + productName);
        System.out.println("Price: $" + price);
    }
}

// Define the Inventory interface
interface Inventory {
    void addProduct(Product product, int quantity);

    void displayInventory();
}

// Define the Sales interface
interface Sales {
    void sellProduct(String productId, int quantity);
}

// Define the InventoryManager class implementing Inventory and Sales interfaces
class InventoryManager implements Inventory, Sales {
    Map<Product, Integer> stock;

    InventoryManager() {
        this.stock = new HashMap<>();
    }

    @Override
    public void addProduct(Product product, int quantity) {
        if (stock.containsKey(product)) {
            int existingQuantity = stock.get(product);
            stock.put(product, existingQuantity + quantity);
        } else {
            stock.put(product, quantity);
        }
        System.out.println("Product added to the inventory!");
    }

    @Override
    public void displayInventory() {
        System.out.println("Current Inventory:");
        for (Map.Entry<Product, Integer> entry : stock.entrySet()) {
            Product product = entry.getKey();
            int quantity = entry.getValue();
            System.out.println("Product: " + product.productName + ", Quantity: " + quantity);
        }
    }

    @Override
    public void sellProduct(String productId, int quantity) {
        for (Map.Entry<Product, Integer> entry : stock.entrySet()) {
            Product product = entry.getKey();
            if (product.productId.equals(productId)) {
                int availableQuantity = entry.getValue();
                if (quantity <= availableQuantity) {
                    stock.put(product, availableQuantity - quantity);
                    System.out.println("Product sold successfully!");
                } else {
                    System.out.println("Insufficient stock to sell.");
                }
                return;
            }
        }
        System.out.println("Product not found in the inventory.");
    }
}

// Main class to run the program
public class InventorySystem {
    public static void main(String[] args) {
        InventoryManager inventoryManager = new InventoryManager();
        Scanner scanner = new Scanner(System.in);

        // Add products to the inventory
        Product laptop = new Product("L001", "Laptop", 999.99);
        Product smartphone = new Product("S001", "Smartphone", 499.99);

        inventoryManager.addProduct(laptop, 10);
        inventoryManager.addProduct(smartphone, 20);

        // Display current inventory
        inventoryManager.displayInventory();

        // Sell products
        inventoryManager.sellProduct("L001", 5);
        inventoryManager.sellProduct("S001", 15);

        // Display updated inventory
        inventoryManager.displayInventory();
    }
}

File: practice-javascript/inventory.js
// Initial inventory
let inventory = [
    { id: 1, name: 'Laptop', quantity: 10, price: 800 },
    { id: 2, name: 'Smartphone', quantity: 20, price: 500 },
    { id: 3, name: 'Tablet', quantity: 15, price: 300 },
];

// Function to add a new product to the inventory
const addProduct = (id, name, quantity, price) => {
    console.log(`Adding ${quantity} ${name}(s) to the inventory...`);
    const newProduct = { id, name, quantity, price };
    inventory.push(newProduct);
    console.log('Product added:', newProduct);
};

// Function to update the quantity of a product in the inventory
const updateQuantity = (productId, newQuantity) => {
    console.log(`Updating quantity of product with ID ${productId}...`);
    const productIndex = inventory.findIndex(product => product.id === productId);
    if (productIndex !== -1) {
        inventory[productIndex].quantity = newQuantity;
        console.log(`Quantity updated to ${newQuantity}`);
    } else {
        console.log('Product not found in inventory');
    }
};

// Function to generate a sales report
const generateSalesReport = () => {
    console.log('Generating sales report...');
    const totalSales = inventory.reduce((total, product) => total + product.quantity * product.price, 0);
    console.log('Total sales:', totalSales);
};

// Add a new product
addProduct(4, 'Printer', 5, 200);

// Update quantity of existing product
updateQuantity(2, 15);

// Generate sales report
generateSalesReport();

// Display the updated inventory
console.log('Updated Inventory:', inventory);